### DISCUSSION TITLE
`ggside` exports and Code Management

### DISCUSSION THREAD

**Discussion by jtlandis** [jtlandis/ggside#52] [Ideas]
With the ongoing development of version `0.3.0`, I am starting to worry that ggside has a heavy amount of exports. Additionally, when considering how complicated the package has become since its advent, it may be smart to remove some of the exports.

Specifically I am considering removing all geom `ggproto` classes from the exports of `ggside` considering that alone they are not helpful as an extendable object. i.e. the `ggside` variants of any `Geom` object will are not able produce a `ggside` subclass if they are placed in the `geom` argument of `ggplot2::layer()` function.

Now that I am considering extending the scales classes as well as possibly the Layout class to have more control. I want to the code base to be as simple as possible so that the overhead of maintenance is minimal.

Currently, the easiest way to subclass a `ggplot` object to a `ggside` object is to add a `ggside_layer` object. In the `ggside v0.2.3`, this is simply subclassing the `LayerInstance` without any extension. I hope to extend the `LayerInstance` in the future, but minimally as the Layer object is not exported by `ggplot2` and is subject to change... something that I have already been burned on.

Alternatively, I am considering writing a constructor for `ggside_layer` which takes a vanilla ggplot object, and modifies the Geom and Stat slots in a new `LayerInstance`. I already have a prototype working for the Geom ggproto objects, so implementing a factory constructor for the Stat variants should mean minimal subclassing of the `LayerInstance` will be needed. 

#### Facet

* The original workhorse in the advent of `ggside` and would have been the only subclass if I thought of a clever way to associate layers with panel types.

#### Geom

* Originally it was subclassed just to place which panel group (x, y, or main) the layer was associated with. This helped inform data mapping with facets
* Information about default aesthetics are consulted. If I want `xsidey` and `ysidex` to be proper scales for end users to use. 

#### Stat

* Have not subclassed yet, but I am now considering it so that I do not need to override `LayerInstance` slots and methods.

#### Coord

* If I recall correctly, this was subclassed in order to support adding theme elements to side plots separately from the entire plot.

#### Guides

* Like with Stat, it is not subclassed yet, but now that `ggplot2 3.5.0` will be using ggproto classes instead of S3, I should consider subclassing them soon.

#### Scales

* scales are complicated, and I am not yet sure of the payout for subclassing these. Right now, users can mix continuous and discrete axis so long as they declare the side axis. 
* If I commit to changing the default aesthetics for the Geoms, then independent scales may be generated automatically.
* Based on how the Layout works and `scale_apply`, any transform provided is applied to all positional scales. I forget if having its own scale fixes this issue.
* The most tacky solution involves temporarily changing the aesthetics of the scale in `transform_df`
* A lot of the code in the past year _works_ but I do not trust it. I'm sure there is a reason to why it is so complicated but I cannot recall the specific reason for some code chunks like `local_vanilla_scale_aes(self)` function, which will temporarily change the scales aesthetics...

In summary, maintaining and adding features to `ggside` is becoming quite unwieldy. Since the project had a stale period, code that was developed previously was never submitted to CRAN. I get the sense that past me was also weary of submitting something too disorganized.

**Comment by jtlandis** [comment]
There was at one point motivation to extend the Layout ggproto object. This motivation stemmed from the fact we subclassed the scales... 

The specific reason in [this issue post](https://github.com/tidyverse/ggplot2/issues/5077) was that all positional scales are assumed to be the same `self$panel_scales_y[[1]]$aesthetics`, whereas in `ggside` the scales object may have different aesthetics mapped to them.

I need to find the use case in which this was an issue

**Comment by jtlandis** [comment]
Recap on what I did over a year ago...

Tying back to the scales, I wanted Geom's default aesthetics to be covered to use their `xside` or `yside` variants so that these data points would not be transformed with the main axis scale's transformations and can potentially be transformed independently. Creating these aesthetics and having these positional scales added automatically is simple enough to do, however an issue arrises in that the `xsidey` and `ysidex` scales do not map the layer data correctly in `layout$map_position` because it is hard coded to only look at `self$panel_scales_y[[1]]$aesthetics`, i.e. the first scale of that position type because they assume the same aesthetics will be there.

I had coded round this by changing `facet$train_scales` which is called by `Layout$train_position`, to modify the available aesthetics to include `xside` or `yside` prefixes such that `Layout$map_position` would work.

This also explains why `local_vanilla_scale_aes(self)` was placed in a subclassed version of positional scales' `$transform_df` function.

Ideally, I just extend the Layer$map_position to consider all aesthetics. This may allow me refert `facet$train_scales` and not subclass the ScalesList object. 🤞
